lib/services/local/users.go

```diff
@@ -73,6 +74,186 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+			if out[len(out)-1].GetName() != prevUser {
+				user, err := userFromUserItems(prevUser, users[prevUser])
+				if err != nil {
+					return nil, "", trace.Wrap(err)
+				}
+
+				out = append(out, user)
+			}
+
+			return out, "", nil
+		}
+
+		// If the number of users retrieved is more than the requested page size
+		// then return the number of users requested and the token for the user
+		// at page size + 1 in the slice.
+		if len(out) > pageSize {
+			nextKey := backend.GetPaginationKey(out[pageSize])
+			// Truncate the last item that was used to determine next row existence.
+			clear(out[pageSize:])
+			return out[:pageSize], nextKey, nil
+		}
+
+		// Since local auth secrets share the same prefix another batch of items from
+		// the backend may be necessary to fulfill the requested page. If the current
+		// start key matches the last item in the results then this page has already been
+		// retried once and means there are no more user entries to search so return the
+		// partial page of users.
+		nextStart := []byte(backend.NextPaginationKey(paginationKeyResource{name: string(result.Items[len(result.Items)-1].Key)}))
+		if bytes.Equal(rangeStart, nextStart) {
+			return out, "", nil
+		}
+
+		rangeStart = nextStart
+	}
+}
+
+// ListUsers returns the next page of users.
+func (s *IdentityService) ListUsers(ctx context.Context, pageSize int, nextToken string, withSecrets bool) ([]types.User, string, error) {
+	if withSecrets {
+		users, next, err := s.listUsersWitSecrets(ctx, pageSize, nextToken)
+		return users, next, trace.Wrap(err)
+	}
+
+	rangeStart := backend.Key(webPrefix, usersPrefix, nextToken)
+	rangeEnd := backend.RangeEnd(backend.ExactKey(webPrefix, usersPrefix))
+
+	// Adjust page size, so it can't be too large.
+	if pageSize <= 0 || pageSize > apidefaults.DefaultChunkSize {
+		pageSize = apidefaults.DefaultChunkSize
+	}
+
+	// Artificially inflate the limit to account for user secrets
+	// which have the same prefix.
+	limit := pageSize * 4
+
+	var out []types.User
+	for {
+		result, err := s.Backend.GetRange(ctx, rangeStart, rangeEnd, limit)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		// If there were no results just return an empty slice.
+		if len(result.Items) == 0 {
+			return out, "", nil
+		}
+
+		for _, item := range result.Items {
+			if !bytes.HasSuffix(item.Key, []byte(paramsPrefix)) {
+				continue
+			}
+
+			resource, err := services.UnmarshalUser(item.Value, services.WithRevision(item.Revision))
+			if err != nil {
+				return nil, "", trace.Wrap(err)
+			}
+
+			// Prevent adding duplicate resources
+			if len(out) != 0 && out[len(out)-1].GetName() == resource.GetName() {
+				continue
+			}
+
+			out = append(out, resource)
+		}
+
+		// If the number of users retrieved is more than the requested page size
+		// then return the number of users requested and the token for the user
+		// at page size + 1 in the slice.
```

Comment:
Aren't we risking an information leak if we return the name of the next user rather than the key after the last returned user? Can anyone call `ListUsers` without having full read on all users anyway?